Add style rule for field order in simple structs with member functions. (#2535)
carbon-language/carbon-lang: docs/project/cpp_style_guide.md

        `namespace Carbon::Testing { namespace { ... } }` to keep everything
        internal.

### Declaration order

-   For a struct without constructors or factory functions, fields should be
    declared at the start of the struct, immediately after any member types or

>     typedefs, not at the end.

Reviewer: Is this the rule as planned? So to be clear, with something like ParenContents having a single factory function, this rule shouldn't apply, right?

Similarly, if we have an existing type:

```
struct Foo {
  int data;

  void Print(...);
}
```

And we decide it reads better with a trivial constructor, then we reorder everything:

```
struct Foo {
  Foo(int data) : data(data) {}

  void Print(...);

  int data;
}
```

This rule seems odd to me, but I can't really object if it's already-approved wording for the Google style guide.

Reviewer: To be clear on what I'd have thought the rule would be, something like "For a struct where the member variables are the main API, member variables should come before methods instead of after in the declaration order."

(but still, I'd mainly be approving on the "already approved" argument)

Author: We can do that, and I think it'll be an improvement, but we'll need to decide on a relative order for constructors/factory functions versus data members. @chandlerc and I have also been discussing a larger change where we'd say that public data members always precede public member functions, and similarly for private data members.

Reviewer: You can of course do that, but I'd find it pretty disorienting to have private data members before private functions. Of course, I already find this disorienting, so 🤷

Reviewer: To be clear though --- I value most matching whatever the Google style guide update is that's planned.

I would assume that's not going to be reordering all data members, though.

Reviewer 2: The change that is being contemplated for the Google style guide would just optionally allow (in structs) public data fields to optionally be placed together in a new position, second after types and type aliases. Whether it is allowed won't hinge on the first factory function or constructor, but whether the type fits the style guide's [definition of a struct vs. a class](https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes).

I've suggested an edit that matches the direction here better I suspect.

----

On the subject of a more significant change...

Google's style guide is somewhat constrained to not make huge amounts of code stop conforming when it changes. As a consequence, the changes there are going to largely focus on allowing more variations rather than providing a precise and prescriptive rule.

FWIW, I can easily imagine a useful precise and prescriptive rule that I would very much like. I would prefer it to the Google style guide not because it significantly better for readability but simply because it would reduce the choices I have to make without too much readability cost, much like I prefer `clang-format` to thinking hard about formatting.

However, I don't think it would easily match Google's style guide which I think has some unfortunate choices that are now hard to remove due to legacy.

Up to you all whether to pursue this or not -- the divergence costs me almost nothing and the more rigid rule would benefit me so I'm fine with it, but without a strong opinion.

Reviewer: > Google's style guide is somewhat constrained to not make huge amounts of code stop conforming when it changes. As a consequence, the changes there are going to largely focus on allowing more variations rather than providing a precise and prescriptive rule.

I understand the friction Google deals with on style changes. I would not have expected the Google style guide to use the wording given, since it seems odd to me in the context of what's being dealt with. I *definitely* don't expect Google to start allowing class variables first due to the friction involved.

But, to be clear, I think there are good reasons to not put class member variables first:

1) Methods are typically more API than variables, even private methods versus private variables.

Typically, if I'm adding a private method to alter a private variable, I don't want that private variable to be modified directly when the method could be used. As a consequence, I would like to impress upon people trying to edit the class what the private methods are before they start looking for variables to directly modify.

2) Putting variables at the end makes them easy to locate if needed.

Classes shouldn't have public variables, per style. As a consequence, variables should always be the last thing in a class. When I'm trying to see what data a class has, it's a quick and consistent place to look.

I don't think there's an equivalence for methods: you're going to find public methods above member variables no matter what. There is no "go to the end to find all methods".

The alternative would mainly be to have multiple private/public sections, then put private variables above everything else; but that conflicts with (1), and you'd need typedefs first anyways.

3) The style _does_ allow private and public methods to be adjacent in many situations.

I think it's fairly common that a class will have public methods actually immediately followed by private methods, so they're all "together" (although private typedefs interrupt this). Constructors and methods also have visually similar signatures, so having constructors next to methods helps.

---

Note, this is also why I'm closer to ambivalent about structs. A struct should have a simple API. Point 1 applies, but not points 2 or 3. But the flipside here is, I think this is a leads decision. Leads seem to be saying to switch style.

> -   For a struct without constructors or factory functions, fields should be
>     declared at the start of the struct, immediately after any member types or
>     typedefs, not at the end.

Reviewer 2: If we want to match the upstream direction, we should use the style guide's existing distinction between a "struct" and a "class" which provides most of the nuance needed here:

```suggestion
-   For a [struct as opposed to a class](https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes), fields should be
    declared at the start of the struct, immediately after any public member types or
    typedefs.
```

Notably, adding constructors or factory functions are all fine. It is only introducing a *type invariant* that would trigger a change, and it would also trigger a change to use the `class` keyword as well as other changes.

Author: My concern here is that if a struct starts as:

```
struct Foo {
  int a;
  std::string b;

  // ... other stuff ...
```

that people will try to initialize it as `Foo{.a = 1, .b = "hello"}` and be surprised if that fails. And it will fail if `Foo` also has a constructor. Similarly if `Foo` has factory functions, people might not use them when appropriate if aggregate initialization is available and the members precede the factory functions. I'd be fine with this ordering:

- member types
- constructors and factory functions
- public data members
- public member functions
- ...

but that's not consistent with the new rule in the baseline style guide.
